input_pipeline/oracle.py

```diff
@@ -0,0 +1,171 @@
+  processed (in the arcs history), the method returns true."""
+  for dependent in gold_dependents[token_id]:
+    if (token_id, dependent) not in arc_history:
+      return False
+  return True
+
+def shortest_stack_oracle(stack: List[int],
+                          buffer: List[int],
+                          arc_history: List[Tuple[int, int]],
+                          gold_labelled_edges: Dict[Tuple[int, int], int],
+                          gold_dependents: Dict[int, List[int]],
+                          sentence: List[int]
+                          ) -> Tuple[int, int]:
+  """
+  Shortest stack oracle. Try to apply a left-arc, then a right-arc, and finally
+  apply a shift transition. TODO: link to paper/explanation.
+
+  Args:
+    stack: transition-system stack. This is where the tree is constructed. The
+      stack is represented by a python list, and the elements are tokens
+      (positions in the input sentence).
+    buffer: initially the input sentence, it's where the non-processed tokens
+      are stored. It's represented by a python list of tokens.
+    arc_history: List of processed arcs.
+    gold_labelled_edges: Output tree -represented as a set of labeled edges,
+      more concretely through a dictionary of (token1, token2) -> arc_label_id.
+    gold_dependents: Dictionary of dependents for each token, that is
+      token id -> list of dependents.
+    sentence: Input sentence.
+  Returns:
+    the gold action for the given configuration or None if no action can be
+    generated.
+  """
+  # make sure there are at least two tokens on the stack.
+  if len(stack) >=2:
+    stack_top = stack[-1]
+    stack_2nd_top = stack[-2]
+    # Try Left-Arc, that is link the top of the stack to the second top of the
+    # stack and pop the second to top of the stack.
+    if (stack_top, stack_2nd_top) in gold_labelled_edges.keys():
+      action_label = gold_labelled_edges[(stack_top, stack_2nd_top)]
+      return (ArcStandardAction.LEFT_ARC, action_label)
+    # Try Right-ARC (check if there is a link from the second to the top of the
+    # stack and that the top has no unprocessed dependents).
+    if (stack_2nd_top, stack_top) in gold_labelled_edges.keys()\
+      and check_dependents(stack_top, gold_dependents, arc_history):
+      action_label = gold_labelled_edges[(stack_2nd_top, stack_top)]
+      return (ArcStandardAction.RIGHT_ARC, action_label)
+  # Try Shift.
+  if len(buffer)>0:
+    shifted_token = sentence[buffer[0]]
+    return (ArcStandardAction.SHIFT, shifted_token)
+  # If no action was returned, return None.
+  return None
+
+def apply_action(action_type: int,
+                 stack: List[int], buffer: List[int],
+                 arc_history: List[Tuple[int, int]]):
+  """Applies the given action, modifying the input lists (stack and buffer)."""
+  if action_type == ArcStandardAction.LEFT_ARC:
+    # Store the arc in arc history.
+    arc_history.append((stack[-1], stack[-2]))
```

Comment:
Maybe use them same as in the `shortest_stack_oracle` function:
    `stack_top = stack[-1]`
    `stack_2nd_top = stack[-2]`